### ISSUE TITLE
🪲 Precision of floating points calculations

### ISSUE THREAD

**Issue by jpelay** [hedyorg/hedy#5819]
A teacher reported this to us in an email:

> 
> Hi,
> 
> Level 12 Calculator - Multiplying Decimals
> 
> When the code is run and the player answers 1.95 the output is 'Wrong!'
> ```
> number1 = 1.3
> number2 = 1.5
> player_answer = ask 'What is ' number1 ' multiplied by ' number2
> correct_answer = number1 * number2
> print 'your answer was ' player_answer
> print number1 ' times ' number2 ' is ' correct_answer
> if player_answer = correct_answer
>     print 'Correct!'
> else
>     print 'Wrong!'
> ```

I checked and the reason is IEE754, because this is the result that Pythons gives:

```
>>> 1.3*1.5
1.9500000000000002
>>> 1.95 == (1.3*1.5)
False
>>>
```

So, two lines of argumentation can be given with respect to how we can deal with things like these:

In pro of keeping this behavior: since Hedy is teaching programming to kids, and the end goal is to end up programming in Python, this behavior should be kept. Ultimately, this is how floating point numbers are going to work in pretty much every programming language. Changing the above comparison to work, would cause confusion if they eventually move to Python and then figuring out that floating point numbers are funky.

Against: We should change it because we are trying to simplify programming, and teaching it step-by-step, and part of that requires simplifying how floating point numbers work, so it's up to students to later on adapt to the realities of IEE754. There must be a library that has floating point numbers that work like the ones we are used to, specially for accounting, where such things are essential.

**Comment by arjank** [comment]
A quick Google search led me to an answer on StackOverflow which pointed to the [Decimal type](https://docs.python.org/3.11/library/decimal.html).

**Comment by jpelay** [comment]
> A quick Google search led me to an answer on StackOverflow which pointed to the [Decimal type](https://docs.python.org/3.11/library/decimal.html).

It's not implemented in Skulpt 😢 so we can't use that :/

**Comment by AnneliesVlaar** [comment]
<img width="1303" alt="Scherm­afbeelding 2024-10-19 om 08 34 01" src="https://github.com/user-attachments/assets/052622c5-2d0f-4b7d-94de-ed89b36cfc6e">

I had a similar (?) problem in level 6. I would expect 360/19 to be 19 not 18.
I agree with 'against keeping this behavior'.

**Comment by FBastiaan04** [comment]
Floating point errors are an infamous problem in computer science. The floating points defined by IEE754 are however still the best way of getting high accuracy in both very large and very small numbers. As such, they are used by almost all programming languages and even physically integrated into the CPU. In my opinion, explaining what they are and that they can cause these kinds of errors is better than having students run into them in the wild on their own. Fully replacing them with a fixed point type (like the decimal type) would also be a *lot* of effort.

The bug that 360/19 returns 18 is different, and not caused by floating point errors. When running this line in python, it returns `18.94736842105263`. This issue is caused by hedy using the default conversion of `float`s into `int`s, which is the floor function (always round down). Python also has a `round` function, but be warned! This function defaults to **gaussian rounding**. A type of rounding which rounds 2.5 to 2 and 3.5 to 4. This is not how students are taught to round in school, and is highly confusing. The best way to solve this is to write our own rounding function, but that's only like 7 lines of code.

In short, I suggest changing our behaviour sothat rounding is done as students expect it. I do however believe that using fixed point instead of floating point would be far too much effort, and also would give our students a big hurdle later on when they start using python or any other programming language and run into IEE754 themselves.

**Comment by jpelay** [comment]
I agree with the comment I dont think this would be a huge change, and would make sense on most cases.
